### PR TITLE
Avoid diff anomaly if the new proof has no foreground goal

### LINKED ISSUE #15578: Proof Diff Anomaly
#### Description of the problem

In Coqtail and CoqIDE the following script prints `Anomaly "Uncaught exception Failure("hd")"` at the indicated line. There is no anomaly if proof diffs are disabled.
Possibly related: #14425

```coq
Set Diffs "on".

Variable A: Type.

Inductive P : list A -> list A -> Prop :=
  | P1 : forall x xs, P xs (x :: xs)
  | P2 : forall x xs ys, P xs ys -> P xs (x :: ys).

Goal well_founded P.
Proof.
  red; enough (forall ys xs, P xs ys -> Acc P xs).
  - intros [].
    + admit.
    + eapply H.
      constructor. (* <- Anomaly "Uncaught exception Failure("hd")." *)
```

#### Coq Version

8.14, master (381b6e5aca8a21f9895784920d70f768d5851099)

### PR DESCRIPTION
**Pull request by jfehrle**
Fixes: #15578

### CODE REVIEW

File: printing/proof_diffs.ml [hunk 1]
```diff
@@ -633,25 +639,25 @@ let make_goal_map op np =
   | [hd] ->
     (* only 1 removal, some additions *)
     Evar.Set.fold (fun x accu -> GoalMap.add x hd accu) add_gs ng_to_og
+  | _ when Evar.Set.is_empty add_gs -> ng_to_og (* only removals *)
```
Comment (ppedrot) [inline printing/proof_diffs.ml:642]: Don't use when clause, this is quite fragile. Rather, inline the if branch inside the default pattern.

Comment (jfehrle) [reply printing/proof_diffs.ml:642]: Why do you consider it fragile?

Comment (ppedrot) [reply printing/proof_diffs.ml:642]: It's not robust w.r.t. completeness checks.

Comment (jfehrle) [reply printing/proof_diffs.ml]: Updated.  @herbelin, @ejgallego, who can be the assignee for this one?

@ppedrot You'd like to see the `when` clause be removed from OCaml?

**Review by ppedrot — APPROVED** [review]

File: printing/proof_diffs.ml (hunk 1 above)
Comment (ppedrot) [reply printing/proof_diffs.ml]: @jfehrle it's the usual tradeoff of the so called expressivity problem, and it depends on the context. I prefer explicit extensibility over implicit extensibility in general, but that doesn't we should remove the ability to do it the other way around.

**Comment by ppedrot** [comment]
I am assigning and merging, this is trivial and there is no need to wait sync from others.

**Comment by ppedrot** [comment]
@coqbot merge now

**Comment by SkySkimmer** [comment]
This doesn't backport cleanly to 8.15. @jfehrle or @ppedrot do you want to make a 8.15-specific PR or shall we leave this fix as 8.16-only?

**Comment by jfehrle** [comment]
I can change it, should take long at all.

**Comment by jfehrle** [comment]
Isn't it already included in v8.15 as 6ad3df958b4a4537b14f12a94edb5c013c96be0f?

**Comment by SkySkimmer** [comment]
That's not in 8.15 AFAICT

**Pull request merged**